vignettes/available-theming-variables.Rmd

```diff
@@ -0,0 +1,249 @@
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  echo = FALSE, 
+  message = FALSE
+)
+
+
+htmltools::tagList(rmarkdown::html_dependency_font_awesome())
+```
+
+```{r setup}
+library(dplyr)
+library(readr)
+library(stringr)
+library(purrr)
+library(glue)
+library(bslib)
+```
+
+```{css}
+table.dataTable tr[role="row"] td {
+  font-family: monospace;
+}
+table.dataTable tr[role="row"] td:nth-child(3) {
+  font-family: inherit;
+  font-size: 0.85rem;
+}
+table.dataTable tr[role="row"] td:nth-child(4) {
+  text-align: center;
+}
+.dep-link {
+  color:steelblue; 
+  text-decoration:underline;
+  cursor: pointer;
+}
+```
+
+```{r variables_df}
+# BS3
+# Not using BS3 now because the comment templates are different so a whole new scraping pipeline is needed
+# sass_variables_loc <- "https://raw.githubusercontent.com/rstudio/bslib/master/inst/lib/bs-sass/assets/stylesheets/bootstrap/_variables.scss"
+
+# BS4
+sass_variables_loc <- rprojroot::find_package_root_file("inst/lib/bs/scss/_variables.scss")
+sass_variables_url <- "https://github.com/rstudio/bslib/blob/master/inst/lib/bs/scss/_variables.scss"
+# A section is denoted by a comment followed by either an empty line or an empty comment.
+# e.g.
+# // Section ID
+#
+# Or
+# // Section ID
+# //
+#
+# Sometimes the sections are just isolated floating comments
+#
+# // Section ID
+#
+#
+
+raw_text <- read_file(sass_variables_loc)
+
+collapse_multiline_comments <- function(lines_df){
+
+  repetitions <- rle(lines_df$is_comment & !lines_df$is_empty_comment)
+  block_id <- rep(seq_along(repetitions$lengths), times = repetitions$lengths)
+  is_comment_block <- rep(repetitions$values, times = repetitions$lengths)
+  # A negative value means it's not a comment block. Using negatives so we
+  # don't match ids of comment blocks
+  lines_df$comment_block <- ifelse(is_comment_block, block_id, -seq_len(nrow(lines_df)))
+
+  group_by(lines_df, comment_block) %>%
+    summarise(
+      across(c(-line), first),
+      line = paste(line, collapse = " ")
+    ) %>%
+    arrange(line_number) %>%
+    select(-comment_block)
+}
+
+link_to_pos_in_script <- function(line_num){
+  paste0(sass_variables_url, "#L", line_num)
+}
+
+tibble(
+  line = strsplit(raw_text, split = "\n")[[1]]
+) %>%
+  mutate(
+    line_number = row_number(),
+    is_comment = str_detect(line, "^\\/\\/"),
+    is_empty_comment = str_detect(line, "^\\/\\/\\s*$")
+  ) %>%
+  collapse_multiline_comments() %>%
+  mutate(
+    below_is_empty_comment = lead(is_empty_comment),
+    above_is_empty_line = lag(line) == "",
+    below_is_empty_line = lead(line) == "",
+    is_section = (is_comment & below_is_empty_comment) | (is_comment & above_is_empty_line & below_is_empty_line),
+    variable = str_extract(line, "^\\$(.+)?(?=:)"),
+    section = ifelse(is_section, str_remove(line, "\\/\\/\\s*"), NA),
+    section_link = ifelse(is_section, link_to_pos_in_script(line_number), NA),
```

Comment:
Sorry for not thinking of this earlier, but instead of rendering everything in one giant table, how about we generate a different table for each section and navigate them with tabset pills? This would make the different sections much easier to navigate. Something like:

```r
tabsetPanel(
  type = "pills",
  tabPanel("Color system", DT::datatableOutput("color_system")),
  ...
)
```

(I'm not sure how easy this would be without Shiny, though, and I'm sure this complicate to the hyperlinking as well since you'd need some sort of map between sections and variables)
 

  should tabs for each section to display a different table

Reply:
Somewhat relatedly, I think it'd prefer scrollable DT table(s) over pagination. Most everyone is not going to know what they're looking for, so the more we can "surface" the content without having to click through it, the better

Reply:
I think a difficulty with the tabs also would be navigating between dependencies. Certainly possible, but a bit more logic to look up what table the dependency lives in, then navigate to that tab, then navigate to the right location

Reply:
As for scrolling DT, I agree. I was using tabs because I couldn't get the scrolling to work properly earlier. I'll take another crack at it today though.

Reply:
I got the scrolling table to work and I think it's way nicer. The dependency finding is much more intuitive with animated scrolling

Reply:
Yes, this is much better, thanks!